sdk/cosmos/azure_data_cosmos/src/clients/container_client.rs

```diff
@@ -0,0 +1,230 @@
+    /// If you want to deserialize the document to a direct representation of the JSON returned, use [`serde_json::Value`] as the target type.
+    ///
+    /// We recommend using "turbofish" syntax (`query_items::<SomeTargetType>(...)`) to specify the target type, as it makes type inference easier.
+    ///
+    /// # Arguments
+    ///
+    /// * `query` - The query to execute.
+    /// * `partition_key` - The partition key to scope the query on.
+    /// * `options` - Optional parameters for the request.
+    ///
+    /// # Examples
+    ///
+    /// The `query` and `partition_key` parameters accept anything that can be transformed [`Into`] their relevant types.
+    /// This allows simple queries without parameters to be expressed easily:
+    ///
+    /// ```rust,no_run
+    /// # async fn doc() {
+    /// # use azure_data_cosmos::{CosmosClient, CosmosClientMethods, clients::DatabaseClientMethods, clients::ContainerClientMethods};
+    /// # let credential = azure_identity::create_default_credential().unwrap();
+    /// # let client = CosmosClient::new("https://myaccount.documents.azure.com/", credential, None).unwrap();
+    /// # let db_client = client.database_client("my_database");
+    /// # let container_client = db_client.container_client("my_container");
+    /// #[derive(serde::Deserialize)]
+    /// struct Customer {
+    ///     id: u64,
+    ///     name: String,
+    /// }
+    /// let items = container_client.query_items::<Customer>(
+    ///     "SELECT * FROM c",
+    ///     "some_partition_key",
+    ///     None).unwrap();
+    /// # }
+    /// ```
+    ///
+    /// You can specify parameters by using [`Query::from()`] and [`Query::with_parameter()`]:
+    ///
+    /// ```rust,no_run
+    /// # async fn doc() {
+    /// # use azure_data_cosmos::{CosmosClient, CosmosClientMethods, clients::DatabaseClientMethods, clients::ContainerClientMethods, Query};
+    /// # let credential = azure_identity::create_default_credential().unwrap();
+    /// # let client = CosmosClient::new("https://myaccount.documents.azure.com/", credential, None).unwrap();
+    /// # let db_client = client.database_client("my_database");
+    /// # let container_client = db_client.container_client("my_container");
+    /// #[derive(serde::Deserialize)]
+    /// struct Customer {
+    ///     id: u64,
+    ///     name: String,
+    /// }
+    /// let query = Query::from("SELECT COUNT(*) FROM c WHERE c.customer_id = @customer_id")
+    ///     .with_parameter("@customer_id", 42).unwrap();
+    /// let items = container_client.query_items::<Customer>(query, "some_partition_key", None).unwrap();
+    /// # }
+    /// ```
+    ///
+    /// See [`PartitionKey`] for more information on how to specify a partition key, and [`Query`] for more information on how to specify a query.
+    fn query_items<T: DeserializeOwned + Send>(
+        &self,
+        query: impl Into<Query>,
+        partition_key: impl Into<PartitionKey>,
```

Comment:
Why did we choose to elevate `partition_key` to be a top level parameter?  In .NET's `Container.GetItemQueryIterator`(https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.container.getitemqueryiterator?view=azure-dotnet) it's part of the `QueryOptions`.

Reply:
The main reason is that we _require_ a partition key. Currently, we are not supporting cross-partition queries, similar to the Go SDK. Therefore, we require a `partition_key`. Also, I modelled this API on [the similar API in the Go SDK](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/data/azcosmos/cosmos_container.go#L491). which requires the partition key be provided.

If we want to support cross-partition queries later, we could probably still use this API. Rust doesn't support overloading, so we have to keep the arguments the same, or introduce a new method. The `PartitionKey` struct this expects contains a `Vec<PartitionKeyValue>` which is expected to contain 1 or more keys (to support hierarchical keys). If you're doing a cross-partition query, we could update `PartitionKey` to allow 0 keys to be provided, and add an impl of `From<()> for PartitionKey` that would allow you to write this:

```rust
container.query_items("SELECT * FROM c", (), None); // The "()" indicates no partition key filter
```

Reply:
If it's required - which given the param type, it is - it should be a parameter. If it's optional it should go into parameters.

.NET also establishes 6 as the max number of parameters (can't remember if inclusive or exclusive) of when all parameters should go into an options bag, though the name of that bag has changed over the years. For Rust, I'd prefer we make the method options required but still called `<MethodName>Options` when that happens. You good with that @JeffreyRichter, @RickWinter?

Reply:
Ok, the next iteration will have a slightly modified approach here. Combining this with my realization that serde treats `()` as `null`, so we can use that to represent null partition keys (which means it can't represent "cross-partition" queries here), I've adjusted my approach a bit.

Instead of `Into<PartitionKey>`, we accept `Into<QueryPartitionStrategy>`. There is only one strategy available today, `SinglePartition`. Any value that is `Into<PartitionKey>` is automatically `Into<QueryPartitionStrategy>` and results in `QueryPartitionStrategy::SinglePartition(...)`.

I did a little bit of API evolution "testing" and adding a second variant, `QueryPartitionStrategy::CrossPartition` would then allow us to call `query_items` like this:

```rust
container.query_items("SELECT * FROM c", QueryPartitionStrategy::CrossPartition, None);
```

While still preserving all these other ways of calling it:

```rust
container.query_items("SELECT * FROM c", (), None); // single "null" partition key, i.e. `[null]`
container.query_items("SELECT * FROM c", "my_key", None); // single partition key, i.e. `["my_key"]`
container.query_items("SELECT * FROM c", ("parent", "child"), None); // hierarchical partition key, i.e. `["parent", "child"]`

let pk = PartitionKey::from(("parent", "child"));
container.query_items("SELECT * FROM c", pk, None); // explicit partition key object
container.query_items("SELECT * FROM c", QueryPartitionStrategy::SinglePartition(pk), None); // explicit partition strategy
```

Reply:
I'm okay with this approach, but I feel like the default approach is for queries to be cross-partition, and this will be a little bit of boiler-plate on all of those.

@kirankumarkolli or @jcocchi - any thoughts on the shape of this API?

Reply:
I'd also be open to enabling support for cross-partition queries now, **limited** to the support the Gateway provides for cross-partition queries (so no aggregations, etc), by just setting the `x-ms-documentdb-query-enablecrosspartition` header. The previous unofficial client [supported that](https://docs.rs/azure_data_cosmos/latest/azure_data_cosmos/prelude/struct.QueryDocumentsBuilder.html#method.query_cross_partition).

If we did that, I'd be OK with moving `partition_key` to the Options and "defaulting" to cross-partition if it's not specified. That's actually what the previous unofficial client versions did

Reply:
The problem with allowing *some* cross partition queries now is validation. IIRC the Gateway doesn't reject query shapes it can't handle, it just returns results that may be incorrect because they require extra client-side processing. We didn't open up limited cross partition query support on Go because we chose query correctness over adding functionality users may not read all the limitations of. However, Go allows users to explicitly set this header themselves to unblock if they need to (presumably after understanding the limitations).

Reply:
I'm going to go ahead with this as-is for now. We will need to do some API review for this later, and can have another opportunity to dive into it there.